base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs

```diff
@@ -1173,7 +1212,9 @@ impl LMDBDatabase {
         header: &BlockHeader,
         body: AggregateBody,
         smt: Arc<RwLock<OutputSmt>>,
+        allow_smt_change: Arc<AtomicBool>,
     ) -> Result<(), ChainStorageError> {
+        let can_we_change_smt = allow_smt_change.load(Ordering::SeqCst);
```

Comment:
There is a lot happening in this function outside of whether we can change the SMT or not, and I don't _think_ it matters for most of the function, as we don't end up updating the SMT until the end.

Is this something that we could not do in this function, and somehow test at the caller after this returns, to see fi we also need to call `insert_smt` ?

Reply:
This is a weird design detail, and I think the least worse one there is. 
Main issue we have is we have the DB which is the source of truth, now we have the memory cache of the SMT because loading it takes too long. Now we want to create a File copy of the cache as calculating it takes too long. 

So the DB needs to stay the source of truth and always needs to be correct and in sync. 

But all other patterns require multiple rwlock's which might lead to an inconsistent DB state, and or requires manual dev implementation to dave the SMT